provider/spark.go

```diff
@@ -245,6 +283,256 @@ func (spark *SparkOfflineStore) RegisterResourceFromSourceTable(id ResourceID, s
 	return nil, nil
 }
 
+func (s *S3Store) ResourcePath(id ResourceID) string {
+	return fmt.Sprintf("s3://%s/%s", s.bucketPath, s.resourcePrefix(id))
+}
+
+func stringifyValue(value interface{}) string {
+	switch reflect.TypeOf(value).String() {
+	case "string":
+		return fmt.Sprintf(`"%s"`, value.(string))
+	case "int":
+		return fmt.Sprintf(`%d`, value.(int))
+	case "float":
+		return fmt.Sprintf(`%f`, value.(float64))
+	case "float32":
+		return fmt.Sprintf(`%f`, value.(float32))
+	case "int32":
+		return fmt.Sprintf(`%d`, value.(int32))
+	case "bool":
+		return fmt.Sprintf(`%t`, value.(bool))
+	case "time.Time":
+		//convert to int64
+		return fmt.Sprintf(`%d`, value.(time.Time).UnixMilli())
+	}
+	return ""
+}
+
+func stringifyStruct(data interface{}) string {
+	curStruct := reflect.ValueOf(data)
+	structType := curStruct.Type()
+	structString := `
+	{`
+	for j := 0; j < curStruct.NumField(); j++ {
+		structString += fmt.Sprintf(`"%s":`, strings.ToLower(structType.Field(j).Name))
+		value := curStruct.Field(j).Interface()
+		structString += stringifyValue(value)
+		if j != curStruct.NumField()-1 {
+			structString += ",\n"
+		} else {
+			structString += "\n}"
+		}
+	}
+	return structString
+}
+
+func stringifyStructArray(data interface{}) ([]string, error) {
+	array := reflect.ValueOf(data)
+	structStringArray := make([]string, array.Len())
+	for i := 0; i < array.Len(); i++ {
+		structStringArray[i] = stringifyStruct(array.Index(i).Interface())
+	}
+	return structStringArray, nil
+}
+
+func stringifyStructField(data interface{}, idx int) string {
+	schemaStruct := reflect.ValueOf(data)
+	typeOfS := schemaStruct.Type()
+	fieldDataType := reflect.TypeOf(schemaStruct.Field(idx).Interface()).String()
+	jsonFriendlyFieldName := strings.ToLower(typeOfS.Field(idx).Name)
+	switch fieldDataType {
+	case "string":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=BYTE_ARRAY, convertedtype=UTF8"}`, jsonFriendlyFieldName)
+	case "int":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=INT32"}`, jsonFriendlyFieldName)
+	case "float":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=FLOAT"}`, jsonFriendlyFieldName)
+	case "float32":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=FLOAT"}`, jsonFriendlyFieldName)
+	case "int32":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=INT32"}`, jsonFriendlyFieldName)
+	case "bool":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=BOOLEAN"}`, jsonFriendlyFieldName)
+	case "time.Time":
+		return fmt.Sprintf(`{"Tag": "name=%s, type=INT64"}`, jsonFriendlyFieldName)
+	}
+	return ""
+}
+
+func parquetSchemaHeader() string {
```

Comment:
unit test

Reply:
should I just make it a variable since it's just a single string return?

Reply:
Yeah that's fine